test/visual-regression/util/calculate-diff.js

```diff
@@ -0,0 +1,90 @@
+const fs = require('fs');
+const resemble = require('node-resemble-js');
+const { getFileNames, createDirectoryIfNotExist } = require('./get-file-names');
+
+const DIFF_THRESHOLD = 0.01;
+
+// A wrapper around fs.readFile to return a promise
+function readFile(fileName) {
+  return new Promise((resolve, reject) => {
+    fs.readFile(fileName, (err, result) => {
+      return err ? reject(err) : resolve(result);
+    });
+  });
+}
+
+// A wrapper around Nightwatch's screenshot function to return a promise
+// Note - Nightwatch's screenshot stores the screenshot only in memory as opposed to saveScreenshot
+function takeScreenshot(browser) {
+  const logScreenshotData = false;
+  return new Promise(resolve => browser.screenshot(logScreenshotData, resolve));
+}
+
+// Compares the values for the baseline image with the screenshot of the browser's current page.
+function executeComparison(baselineImageBuffer, screenshotResult) {
+  const baselineImage = new Buffer(baselineImageBuffer, 'base64');
+  const screenshot = new Buffer(screenshotResult.value, 'base64');
+
+  // Execute ResembleJS to compare the images
+  return new Promise(resolve => {
+    resemble(baselineImage)
+      .compareTo(screenshot)
+      .onComplete(resolve);
+  });
+}
+
+// Writes the diff image file used in the event the baseline doesn't match the screenshot of the browser's current page.
+function createDiffImage(diffFileName, comparisonResult) {
+  const diffImageStream = comparisonResult.getDiffImage();
+  const writer = fs.createWriteStream(diffFileName);
+
+  // A wrapper around the stream to return a promise
+  return new Promise(resolve => {
+    diffImageStream
+      .pack()
+      .pipe(writer)
+      .once('finish', resolve);
+  });
+}
+
+// After executing the comparison operation, inspect the result object to create a diff image and run the test.
+function computeComparisonResult(browser, route, diffFileName, comparisonResult) {
+  const misMatchPercentage = parseFloat(comparisonResult.misMatchPercentage);
+  const changesExceedThreshold = misMatchPercentage > DIFF_THRESHOLD;
+
+  // Execution the test assertion
+  browser.verify.ok(!misMatchPercentage, route);
```

Comment:
Reading the documentation, it looks like this is a wrapper around [Node's assert](https://nodejs.org/api/assert.html#assert_assert_ok_value_message), right?

I think I'm misunderstanding something here, though. It looks like this passes only if the images are mismatched. What am I missing?

Reply:
It is, but if it fails an assert, the execution would stop. With verify, it continues. 

But yeah, wow, that's what I get for "cleaning up" too much. I meant to put `changesExceedThreshold` there. So, the`misMatchPercentage` is how many pixels don't match between images expressed as a percentage, so a value of 100 would mean that not a single pixel matches between images. A value of zero means the screenshots are identical.  The tests were still working as I expected because in all of my tests, the screenshots were either identical or off by more than the `DIFF_THRESHOLD`. So `misMatchPercentage` was zero because they were identical, and therefore falsy, which would mean it would always be the same as `misMatchPercentage` in the assertion. 

Confusing explanation, but yeah I need to change that to `changesExceedThreshold`!

Reply:
Aha! Okay, that makes a _lot_ more sense now! 😄